gravity.sh

```diff
@@ -1,476 +1,626 @@
+      lastLine=$(tail -1 "${piholeDir}/${matterAndLight}")
+      [[ "${#lastLine}" -gt 0 ]] && echo "" >> "${piholeDir}/${matterAndLight}"
     fi
   done
 
   echo -e "${OVER}  ${TICK} ${str}"
 }
 
-gravity_Blacklist() {
-  # Append blacklist entries to eventHorizon if they exist
-  if [[ -f "${blacklistFile}" ]]; then
-    numBlacklisted=$(wc -l < "${blacklistFile}")
-    plural=; [[ "$numBlacklisted" != "1" ]] && plural=s
-    local str="Exact blocked domain${plural}: $numBlacklisted"
-    echo -e "  ${INFO} ${str}"
-  else
-    echo -e "  ${INFO} Nothing to blacklist!"
-  fi
+# Parse unfiltered consolidated blocklist into filtered domains-only format
+gravity_Filter() {
+  local str num
+
+  str="Extracting domains from blocklists"
+  echo -ne "  ${INFO} ${str}..."
+
+  # Parse into hosts file
+  gravity_ParseFileIntoDomains "${piholeDir}/${matterAndLight}" "${piholeDir}/${parsedMatter}"
+
+  # Format file line count as currency
+  num=$(printf "%'.0f" "$(wc -l < "${piholeDir}/${parsedMatter}")")
+  echo -e "${OVER}  ${TICK} ${str}
+  ${INFO} ${COL_LIGHT_BLUE}${num}${COL_NC} domains being pulled in by gravity"
+
+  gravity_Unique
 }
 
-gravity_Wildcard() {
-  # Return number of wildcards in output - don't actually handle wildcards
-  if [[ -f "${wildcardlist}" ]]; then
-    numWildcards=$(grep -c ^ "${wildcardlist}")
-    if [[ -n "${IPV4_ADDRESS}" ]] && [[ -n "${IPV6_ADDRESS}" ]];then
-      let numWildcards/=2
-    fi
-    plural=; [[ "$numWildcards" != "1" ]] && plural=s
-    echo -e "  ${INFO} Wildcard blocked domain${plural}: $numWildcards"
-  else
-    echo -e "  ${INFO} No wildcards used!"
-  fi
+# Sort and remove duplicate blacklisted domains
+gravity_Unique() {
+  local str num
+
+  str="Removing duplicate domains"
+  echo -ne "  ${INFO} ${str}..."
+  sort -u "${piholeDir}/${parsedMatter}" > "${piholeDir}/${preEventHorizon}"
+  echo -e "${OVER}  ${TICK} ${str}"
 
+  # Format file line count as currency
+  num=$(printf "%'.0f" "$(wc -l < "${piholeDir}/${preEventHorizon}")")
+  echo -e "  ${INFO} ${COL_LIGHT_BLUE}${num}${COL_NC} unique domains trapped in the Event Horizon"
 }
 
-gravity_Whitelist() {
+# Whitelist blocklist domain sources
+gravity_WhitelistBLD() {
+  local plural str uniqDomains
+
   echo ""
-  # Prevent our sources from being pulled into the hole
-  plural=; [[ "${sources[@]}" != "1" ]] && plural=s
-  local str="Adding adlist source${plural} to the whitelist"
+  plural=; [[ "${#sources[*]}" != "1" ]] && plural=s
+  str="Adding blocklist source${plural} to the whitelist"
   echo -ne "  ${INFO} ${str}..."
 
-  urls=()
-  for url in "${sources[@]}"; do
-    tmp=$(awk -F '/' '{print $3}' <<< "${url}")
-    urls=("${urls[@]}" ${tmp})
-  done
+  # Create array of unique $sourceDomains
+  # Disable SC2046 as quoting will only return first domain
+  # shellcheck disable=SC2046
+  read -r -a uniqDomains <<< $(awk '{ if(!a[$1]++) { print $1 } }' <<< "$(printf '%s\n' "${sourceDomains[@]}")")
```

Comment:
Only one domain is read from ${sourceDomains[@]} at this point. Tried removing the quotes, doesn't appear to fix it.

Reply:
I've pushed a few changes, but I haven't fixed this up yet - I'm able to reproduce the issue now though!